### ISSUE TITLE
Help with PR approach for component validation

### ISSUE THREAD

**Issue by eavonius** [roots/acorn#302]
### Summary

Background: I wanted to use laravel validation in blade components within an acorn project (a sage 10 theme to be specific). I read a bunch of the acorn and laravel 9 source and got it working. However, I'm not sure the best way to contribute this back to acorn, since it looks like people were working on validation already and I don't want to disrupt that if my approach conflicts at all. So I'm Looking for some help with how I should approach potentially contributing this, or helping validation move along!

First I found the file [src/Illuminate/Foundation/Validation/ValidationRequests.php](https://github.com/roots/acorn/blob/3c557618f860471977c94a5cf3eae0d6bc1b21e0/src/Illuminate/Foundation/Validation/ValidatesRequests.php) in the acorn source. This is a PHP trait that can be added to classes in an acorn project that are instantiated with access to the dependency injection container. Composers and Components in a sage 10 theme fit this bill.

Next I added this trait by importing it into a Component class with ```use Illuminate\Foundation\Validation\ValidatesRequests``` then adding a call to ```use ValidatesRequests;``` before the constructor of the class. 

Finally I added some code to the constructor of the component class to test it:

```php
public function __construct(Request $request) {

  if ($request->isMethod('post')) {

    try {
      $this->validate($request, [
        'name' => 'required|min:2'
      ]);
    } catch (\Exception $e) {
      error_log("JSON message dump of validation errors: " . $e->validator->messages());
    }
  }
}
```

When I submit a form on the component that POSTs an input field named "name", and set that field to 1 character long before posting (so the validation fails), the validation exception is thrown but the JSON representation of the message just had the path within an expected localization file to get the actual message to display. E.g:

```json
{"name":["validation.min.string"]}
```

After a bunch of digging around, I found [this post on StackOverflow](https://stackoverflow.com/a/76173064) that showed in the more recent versions of laravel, the ```trans()``` function must be passed to the constructor of the ValidationFactory when it is instantiated. Additionally, the validation.php file from laravel must be present at the default path (relative to your theme/project directory) of ```./lang/en/validation.php```.

So I first copied [the latest commit from laravel's source for version 9.x of validation.php](https://github.com/laravel/framework/blob/d50a2621e8ea9802d7aa4adaa07d856d3a634830/src/Illuminate/Translation/lang/en/validation.php) to that path in my project.

Next I found that on [line 98 of acorn's ValidationRequest.php trait file](https://github.com/roots/acorn/blob/3c557618f860471977c94a5cf3eae0d6bc1b21e0/src/Illuminate/Foundation/Validation/ValidatesRequests.php#L98) is where the ```ValidationFactory``` is retrieved from the container. I just added a second parameter to the ```app``` function, to pass the result of the ```trans``` function as described in the StackOverflow post (since this seemed to take additional parameters to pass to the constructor):

```php
return app(Factory::class, [trans()]);
``` 

After submitting the form. The messages come back localized! 

```json
{"name":["The category name must be at least 2 characters."]}
```

So it looks like that class would need to be modified in acorn. But it also seems like we would want to include instructions in acorn (and possibly update the sage 10 new project template files) to include laravel 9's default validation localization files.

Anyway, happy to answer any other questions if I'm headed in the right direction with this. If not, let me know what other approach I should be taking. Thanks!

**Comment by eavonius** [comment]
Interesting. I tried using the Validator facade like that first, I kept getting errors that no object named "validator" was registered in the container. It seems the factory needs to already be registered with the container for that to work.

I also forgot to mention, I had to add a reference to two other laravel packages for this to work:

"illuminate/validation": "v9.52.14",
"illuminate/translation": "v9.52.14"

**Comment by eavonius** [comment]
Thanks for the info about changes only being in acorn - this is easier actually than I thought.

All that's needed is to add those two references to composer and to add the localization file. The code change to pass the ```trans()``` function isn't needed. I just happened to do it at the same time and didn't realize it was unecessary.

**Comment by Log1x** [comment]
Strange, it looks like the validation service provider is supposed to [register the factory with the container](https://github.com/laravel/framework/blob/d3586ebfd88d1ac62035289d1172b7978cc8e429/src/Illuminate/Validation/ValidationServiceProvider.php#L24-L43).

Make sure you sanity run `wp acorn optimize:clear` after installing `illuminate/validation` and `illuminate/translation` to ensure the provider is getting loaded. Otherwise, might have to figure out what else needs registered.

**Comment by eavonius** [comment]
Yeah I saw that code too. It doesn't seem to be working properly for whatever reason.

**Comment by Log1x** [comment]
> I did run optimize:clear, not sure what that's supposed to do in this case

It was just to make sure the providers weren't cached causing the new packages to not get loaded.

It's possible it is getting caught up when [getting the translator](https://github.com/laravel/framework/blob/d3586ebfd88d1ac62035289d1172b7978cc8e429/src/Illuminate/Validation/ValidationServiceProvider.php#L32) during registration – will require some tinkering.

**Comment by eavonius** [comment]
No problem, thanks for helping out with this. It's working great actually now that I have the translation in the right place and those two references. I'll be sure to get rid of those if you are able to figure out what's going on there. No rush.

**Comment by Log1x** [comment]
This should now work out of the box in v4 with the routing updates.
